ouroboros-consensus/test-consensus/Test/Dynamic/General.hs

```diff
@@ -356,3 +359,83 @@ prop_general k TestConfig{numSlots, nodeJoinPlan, nodeTopology} mbSchedule
         Map.toAscList $
         fmap Map.toAscList $
         testOutputTipBlockNos
+
+    -- In the paper <https://eprint.iacr.org/2017/573/20171115:00183>, a
+    -- /message/ carries a chain from one party to another. When a party forges
+    -- a block, it \"diffuses\" the chain with that block as its head by
+    -- sending a message to each other party (actually, to itself too, but
+    -- that's ultimately redundant). The adversary is able to delay each
+    -- message differently, so some parties may receive it before others do.
+    -- Once a party receives a message, the party can consider that chain for
+    -- selection.
+    --
+    -- In the implementation, on the other hand, our messages are varied and
+    -- much more granular than a whole chain. We therefore observe a delay
+    -- analogous to the paper's /message/ /delay/ by comparing the slot in
+    -- which a block is added to each node's ChainDB against the slot in which
+    -- that block was forged.
+    --
+    -- Since our mock network currently introduces only negligible latency
+    -- compared to the slot duration, we generally expect all messages to have
+    -- no delay: they should arrive to all nodes during the same slot in which
+    -- they were forged. However, some delays are expected, due to nodes
+    -- joining late and also due to the practicality of the ChainSync and
+    -- BlockFetch policies, which try to avoid /unnecessary/ header/block
+    -- fetches. See the relevant comments below.
+    --
+    -- NOTE: This current property does not check for interminable message
+    -- delay: i.e. for blocks that were never added to some ChainDBs. It only
+    -- checks the slot difference once a message does arrive. This seems
+    -- acceptable: if there are no Common Prefix or Chain Growth violations,
+    -- then each message must have either arrived or ultimately been
+    -- irrelevant.
+    --
+    prop_no_unexpected_message_delays :: Property
+    prop_no_unexpected_message_delays =
+        conjoin $
+        [ prop1 nid s p bno
+        | (nid, m)   <- Map.toList adds
+        , (s, pbnos) <- Map.toList m
+        , (p, bno)   <- Set.toList pbnos
+        ]
+      where
+        -- INVARIANT: these AddBlock events are *not* for EBBs
+        adds = nodeOutputAdds <$> testOutputNodes
+
+        prop1 nid s p bno =
+            counterexample msg $
+            delayOK || noDelay
```

Comment:
Would it be helpful to use `Property` (`===` and `.||.`) instead of `Bool` here, so the operands of the failing equality are printed?

Reply:
I push a refactoring of the property: it cleaned up the point/slot contention and improved the counterexample message. Since one of the comparands is computed, I don't think `===` will be very helpful alone, so I'm just committing to the counterexample.

Reply:
BTW, I just reread this property and I find it more readable :+1: